modules/qualitycontrol/QualityController.ts

```diff
@@ -0,0 +1,366 @@
+import { getLogger } from '@jitsi/logger';
+
+import JitsiConference from "../../JitsiConference";
+import { JitsiConferenceEvents } from "../../JitsiConferenceEvents";
+import { CodecMimeType } from "../../service/RTC/CodecMimeType";
+import RTCEvents from "../../service/RTC/RTCEvents";
+import JitsiLocalTrack from "../RTC/JitsiLocalTrack";
+import RTC from "../RTC/RTC";
+import TraceablePeerConnection from "../RTC/TraceablePeerConnection";
+import JingleSessionPC from "../xmpp/JingleSessionPC";
+import { CodecSelection } from "./CodecSelection";
+import ReceiveVideoController from "./ReceiveVideoController";
+import SendVideoController from "./SendVideoController";
+import { VIDEO_CODECS_BY_COMPLEXITY, VIDEO_QUALITY_LEVELS } from '../../service/RTC/StandardVideoSettings';
+
+const logger = getLogger(__filename);
+const LIMITED_BY_CPU_TIMEOUT = 60000;
+const LAST_N_THRESHOLD = 3;
+
+enum QualityLimitationReason {
+    BANDWIDTH = 'bandwidth',
+    CPU = 'cpu',
+    NONE = 'none'
+};
+
+interface IResolution {
+    height: number;
+    width: number;
+}
+
+interface IOutboundRtpStats {
+    codec: CodecMimeType;
+    encodeTime: number;
+    qualityLimitationReason: QualityLimitationReason;
+    resolution: IResolution;
+    timestamp: number;
+}
+
+interface ISourceStats {
+    avgEncodeTime: number;
+    codec: CodecMimeType;
+    encodeResolution: number;
+    localTrack: JitsiLocalTrack;
+    qualityLimitationReason: QualityLimitationReason;
+    timestamp: number;
+};
+
+interface ITrackStats {
+    encodeResolution: number
+    encodeTime: number;
+    qualityLimitationReason: QualityLimitationReason;
+}
+
+interface IVideoConstraints {
+    maxHeight: number;
+    sourceName: string;
+}
+
+class FixedSizeArray {
```

Comment:
Why is this necessary? You can achieve the same with `Object.seal`:

```
// create array with 42 empty slots
let a = new Array(42);

// fill array with some value because
// empty slots can not be changed after calling Object.seal
a.fill(undefined);

Object.seal(a);
// now a is a fixed-size array with mutable entries
```